File: go/530.go
package main

import "math"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var ret int
var prev *TreeNode

func min(a int, b int) int {
	if a < b {
		return a
	}
	return b
}

func helper(cur *TreeNode) {
	if cur == nil {
		return
	}
	helper(cur.Left)
	if prev != nil {
		ret = min(ret, int(math.Abs(float64(cur.Val-prev.Val))))
	}
	prev = cur
	helper(cur.Right)
}

func getMinimumDifference(root *TreeNode) int {
	ret = 1 << 31
	prev = nil
	helper(root)
	return ret
}

File: cpp/530.cpp
#include <algorithm>
#include <climits>
#include <cstdlib>
#include <iostream>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  int ret = INT_MAX;
  TreeNode *prev;
  void inorder(TreeNode *cur) {
    if (cur == nullptr)
      return;
    inorder(cur->left);
    if (prev != nullptr)
      ret = min(ret, abs(cur->val - prev->val));
    prev = cur;
    inorder(cur->right);
  }
  int getMinimumDifference(TreeNode *root) {
    prev = nullptr;
    inorder(root);
    return ret;
  }
};
